### PR TITLE
Implement orphan file deletion in the cloud-config executor script

### PR DESCRIPTION
**Pull request by dimitar-kostadinov**
**How to categorize this PR?**

/area control-plane
/area os
/kind enhancement

**What this PR does / why we need it**:

Currently, if a set of files is specified in `extensions.gardener.cloud/v1alpha1.OperatingSystemConfig` resource and within the next reconciliation some of them no longer persist in `OperatingSystemConfig`, there is no mechanism to delete them and they will remain on the Node VM. With this PR such mechanism that deletes orphan files is implemented in cloud-config executor script.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener-extension-registry-cache/issues/3

**Special notes for your reviewer**:
N/A

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by rfranzke** [comment]
FYI @majst01 @Gerrit91 @vknabel

**Review by acumino — COMMENTED** [review]
/lgtm

File: pkg/apis/extensions/v1alpha1/types_operatingsystemconfig.go [hunk 1]
```diff
@@ -191,6 +191,10 @@ type OperatingSystemConfigStatus struct {
 	// restarted when a new version has been downloaded.
 	// +optional
 	Units []string `json:"units,omitempty"`
+	// Files is a list of file path that are part of the generated Cloud Config and shall be
```
Comment (rfranzke) [inline pkg/apis/extensions/v1alpha1/types_operatingsystemconfig.go]:
```suggestion
	// Files is a list of file paths that are part of the generated Cloud Config and shall be
```

File: pkg/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh [hunk 2]
```diff
@@ -225,6 +238,12 @@ if [[ -f "{{ .pathKubeletKubeconfigReal }}" ]]; then
   fi
 fi
 
+cat << 'EOF' > "$PATH_CLOUDCONFIG_FILES"
+{{ .cloudConfigFiles }}
+EOF
+
+deleteOrphan
```
Comment (rfranzke) [inline pkg/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh]: Shouldn't this also be moved below into the `if ! diff` condition, i.e. only be executed when the cloud config changes? Why did you move it here? Probably this should only run after the units have been restarted to prevent that we delete files they depend on?

File: pkg/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh [hunk 3]
```diff
@@ -17,6 +19,17 @@ PATH_HYPERKUBE_IMAGE_USED_FOR_LAST_COPY_KUBECTL="{{ .pathHyperKubeImageUsedForLa
 
 mkdir -p "{{ .pathKubeletDirectory }}" "$PATH_HYPERKUBE_DOWNLOADS"
 
+deleteOrphan() {
```
Comment (rfranzke) [inline pkg/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh]:
```suggestion
delete_orphan_files() {
```

File: pkg/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh [hunk 4]
```diff
@@ -17,6 +19,17 @@ PATH_HYPERKUBE_IMAGE_USED_FOR_LAST_COPY_KUBECTL="{{ .pathHyperKubeImageUsedForLa
...
+  if [[ ! -f "$PATH_CLOUDCONFIG_FILES_OLD" ]]; then
+    touch "$PATH_CLOUDCONFIG_FILES_OLD"
+  fi
+  # sort for join to work
+  sort -o "$PATH_CLOUDCONFIG_FILES" "$PATH_CLOUDCONFIG_FILES"
+  # calculates old_files - new_files and deletes them
+  join -t $'\n' -v 1 "$PATH_CLOUDCONFIG_FILES_OLD" "$PATH_CLOUDCONFIG_FILES" | xargs -L1 rm -f
+  mv "$PATH_CLOUDCONFIG_FILES" "$PATH_CLOUDCONFIG_FILES_OLD"
+}
```
Comment (ScheererJ) [inline pkg/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh]: Not sure about our general logging policy in this context, but would it not make sense to at least log what we delete? Unless we expect to delete huge number of files it might be helpful in case for whatever reason something got deleted that should not have.

**Review by rfranzke — COMMENTED** [review]
/lgtm

File: pkg/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh (hunk 2 above)
Comment (dimitar-kostadinov) [reply pkg/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh]: It make sense to perform this cleanup only on cloud config changes. Thanks!

**Review by ScheererJ — APPROVED** [review]
/lgtm
